crates/exex/src/exex.rs

```diff
@@ -0,0 +1,306 @@
+
+        // Process each block and its transactions.
+        for (block, txs, _) in blocks {
+            // Execute the block and handle the result.
+            match execute_block(&mut self.db, block, txs).await {
+                Ok((block, bundle, _, _)) => {
+                    // Seal the block and insert it into the database.
+                    let block = block.seal_slow();
+                    self.db.insert_block_with_bundle(&block, bundle)?;
+                    info!(
+                        block_hash = %block.hash(),
+                        transactions = block.body.len(),
+                        "Block submitted, executed and inserted into database"
+                    );
+                }
+                Err(err) => {
+                    // Log an error if the block execution fails.
+                    error!(
+                        %err,
+                        block_hash = %block.hash(),
+                        transactions = block.body.len(),
+                        "Failed to execute block"
+                    );
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use reth_execution_types::{Chain, ExecutionOutcome};
+    use reth_exex_test_utils::{test_exex_context, PollOnce};
+    use reth_primitives::{
+        address, constants::ETH_TO_WEI, hex, Bytes, Header, Receipt, Receipts, SealedBlock,
+        SealedBlockWithSenders, TransactionSigned, TxEip1559, TxType, B256, U256,
+    };
+    use reth_revm::primitives::AccountInfo;
+    use std::{future::Future, pin::pin, str::FromStr};
+
+    /// The initialization logic of the ExEx is just an async function.
+    ///
+    /// During initialization you can wait for resources you need to be up for the ExEx to function,
+    /// like a database connection.
+    async fn exex_init<Node: FullNodeComponents>(
+        ctx: ExExContext<Node>,
+    ) -> eyre::Result<impl Future<Output = eyre::Result<()>>> {
+        // Open the SQLite database connection.
+        let connection = Connection::open(DATABASE_PATH)?;
+
+        // Initialize the database with the connection.
+        let db = Database::new(connection)?;
+
+        // Create a sender address for testing
+        //
+        // We want to fund artificially this address with some ETH because this is the address used
+        // to send funds in the test transaction.
+        let sender_address = address!("6a3cA5811d2c185E6e441cEFa771824fb355f9Ec");
+
+        // Deposit some ETH to the sender and insert it into database
+        db.upsert_account(sender_address, |_| {
+            Ok(AccountInfo { balance: U256::from(ETH_TO_WEI), nonce: 0, ..Default::default() })
+        })?;
+
+        // Create the Kakarot Rollup chain instance and start processing chain state notifications.
+        Ok(KakarotRollup { ctx, db }.start())
+    }
+
+    #[tokio::test]
+    async fn test_exex() -> eyre::Result<()> {
```

Comment:
@greged93 This test is the most important part to test the logic:
1. I create a exex context
2. I take a random transaction and block from etherscan
3. I fund the sender account before initializing the exex
4. I send artificially to exex a notification with my block to mimic real life situation
5. I reexecute my block and the transactions inside it (the evm configuration here will be the part to be replaced by our evm logic). The rest still is the same in my opinion.
6. At the end I check in the database that the balances of the accounts are correct + the block inserted in the db is the right one with all the information needed.